KAFKA-15541: Add oldest-iterator-open-since-ms metric

nicktelford opened the pull request:
Part of [KIP-989](https://cwiki.apache.org/confluence/x/9KCzDw)

This new `StateStore` metric tracks the timestamp that the oldest surviving Iterator was created.

This timestamp should continue to climb, and closely track the current time, as old iterators are closed and new ones created. If the timestamp remains very low (i.e. old), that suggests an Iterator has leaked, which should enable users to isolate the affected store.

It will report no data when there are no currently open Iterators.

nicktelford committed:
KAFKA-15541: Add oldest-iterator-open-since-ms metric

Part of [KIP-989](https://cwiki.apache.org/confluence/x/9KCzDw).

This new `StateStore` metric tracks the timestamp that the oldest
surviving Iterator was created.

This timestamp should continue to climb, and closely track the current
time, as old iterators are closed and new ones created. If the timestamp
remains very low (i.e. old), that suggests an Iterator has leaked, which
should enable users to isolate the affected store.

It will report no data when there are no currently open Iterators.

nicktelford commented:
@mjsax This is the final PR to add the KIP-989 metrics. Although I'm considering opening two, much smaller PRs:

1. To add some additional tests to ensure we cover all the Iterator constructors
2. To eliminate some redundancy now all the metrics have landed. Notably, we can replace the `AtomicInteger` used by `num-open-iterators` with just counting the cardinality of the `Set<MeteredIterator>` that we maintain for `oldest-iterator-open-since-ms`.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java:
```diff
@@ -169,6 +172,10 @@ private void registerMetrics() {
                 (config, now) -> numOpenIterators.get());
+        StateStoreMetrics.addOldestOpenIteratorGauge(taskId.toString(), metricsScope, name(), streamsMetrics,
+                (config, now) -> openIterators.isEmpty() ? null :
+                    openIterators.stream().mapToLong(MeteredIterator::startTimestamp).min().getAsLong()
```
I don't want to over-engineer (given that we can safely assume that the `openIterator` set should be small), but wondering if this is the best implementation?

In the end, we only want to track the create ts, not the iterators themselves. And for create ts we could just maintain a list if longs, and we would `return list.first()` here, and always append to the end of the list when a new iterator is created? Only "remove" would be more expensive, but we could use a sorted tree for the list, and thus remove would be O(log n) not O(n)).

For this case, we also don't need the `MeteredIterator` helper interface.

Thoughts?

nicktelford replied:
Yeah, I considered something like this, but I was concerned that `remove` could be a problem, in particular for two Iterators with the same timestamp.

~An alternative would be to use a `ConcurrentSkipListMap<KeyValueIterator<?, ?>, Long>`, which would be O(log(n)) for both insert and remove, and eliminate any concerns about duplicate timestamps; whilst still being sorted by timestamp (given the appropriate comparator on construction).~

~This also solves the problem that this structure should probably be thread-safe, because Iterators may be opened/closed from various Interactive Query threads.~

~Unfortunately, in Java 8 (the minimum currently supported version), `ConcurrentSkipListMap#size()` is O(n), and considered "inaccurate". This wasn't addressed until JDK 10, so we will need to keep the `AtomicInteger` for tracking the `num-open-iterators`, for the time-being.~

Actually, this wouldn't work because the `Comparator` used by `ConcurrentSkipListMap` operates on the key, not the value.

:thinking: 

Using `ConcurrentSkipListSet` instead of `HashSet` would give us O(1) insert/remove _and_ let us use `openIterators.first().startTimestamp()` to generate the metric, which is also O(1). It also gives us thread-safety.

Unfortunately, it doesn't let us eliminate the `MeteredIterator`.

Any better ideas?

nicktelford replied:
I've pushed a commit with the switch to `ConcurrentSkipListSet`, but I can change it again if we come up with something better.

mjsax replied:
> in particular for two Iterators with the same timestamp.

Yes. We something that allows for duplicates... Not sure if there is any Java standard library impl... Some people implement a custom multi-set on top of tree-map: https://stackoverflow.com/questions/12565587/does-java-have-a-multiset-data-structure-like-the-one-in-c-stl (not sure if it better?) -- seems to be similar to what you propose with `ConcurrentSkipListSet`?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/metrics/StateStoreMetrics.java:
```diff
@@ -451,6 +455,23 @@ public static void addNumOpenIteratorsGauge(final String taskId,
+                RecordingLevel.INFO,
+                oldestOpenIteratorGauge
+        );
+
```
nit: remove blank line

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStoreTest.java:
```diff
@@ -490,6 +490,31 @@ public void shouldTimeIteratorDuration() {
+            try (final KeyValueIterator<String, String> second = metered.all()) {
+                assertThat((Long) oldestIteratorTimestampMetric.metricValue(), equalTo(oldestTimestamp));
+                mockTime.sleep(100);
+            }
```
It would be better to not close the second iterator here, but close the first one first, to see if the metric advances to the second's iterator create ts -- would need some rewriting of the test; try-with-resource won't allow for proper nesting, but we can still use try-finally.

Might actually be best, to open like 5 iterators and close them in some non-linear order (including closing the oldest one like 2 or 3 times) to verify correct behavior.

nicktelford replied:
I've updated the tests with your first suggestion. I didn't go further as I feel that this case is already adequately tested.

mjsax reviewed (commented):
A few questions/suggestions.

Doing follow-up PRs sounds great! We can merge these ofter feature freeze too (we can always add more tests :))

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStoreTest.java:
```diff
@@ -490,6 +490,37 @@ public void shouldTimeIteratorDuration() {
+
+        // now that the first iterator is closed, check that the timestamp has advanced to the still open second iterator
+        assertThat((Long) oldestIteratorTimestampMetric.metricValue(), equalTo(secondTimestamp));
+        second.close();
```
This `close()` call should go into a `finally` block IMHO:
```
KeyValueIterator<String, String> second = null;
final long secondTimestamp;
try {
    try (final KeyValueIterator<String, String> first = metered.all()) {
        ...
    }
    
    // now that the first iterator is closed, check that the timestamp has advanced to the still open second iterator
    assertThat((Long) oldestIteratorTimestampMetric.metricValue(), equalTo(secondTimestamp));
} finally {
    if (second != null) {
        second.close();
    }
}

nicktelford replied:
Done. I assumed you were suggesting this in all the tests, so I updated them all.

mjsax reviewed (commented):
I think it would be good to get rid of `MeteredIterator`, but also don't see a good solution w/o over-engineering...

The current `ConcurrentSkipListSet` is for sure an improvement over the original version. Happy to merge this. Just a nit on the test code.

nicktelford commented:
@mjsax Not sure why the build is failing. `:streams:compileJava` works fine locally. Perhaps the build got confused, can we repeat it?

mjsax commented:
```
[2024-05-28T11:28:17.911Z] > Task :streams:compileJava
[2024-05-28T11:28:17.911Z] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-16041/streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredMultiVersionedKeyQueryIterator.java:54: warning: [this-escape] possible 'this' escape before subclass is fully initialized
[2024-05-28T11:28:17.911Z]         openIterators.add(this);
[2024-05-28T11:28:17.911Z]                           ^
[2024-05-28T11:28:17.911Z] error: warnings found and -Werror specified
```

I re-triggered, but don't expect it to go away... Maybe Java 21 has some sticter/smarter checks?

nicktelford committed:
Fix visibility of MeteredMultiVersionedKeyQueryIterator

Being public causes a compile warning when we add instances to the
`openIterators` `Set`.

It's not necessary for this class to be public, and infact is probably a
bug, because users should not be constructing instances of this class
themselves.

mjsax commented:
Thanks for the KIP @nicktelford -- this will be super useful in the future!
